File: sample/java/src/main/java/com/datadog/android/sample/user/UserFragment.java
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-2020 Datadog, Inc.
 */

package com.datadog.android.sample.user;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import com.datadog.android.Datadog;
import com.datadog.android.log.Logger;
import com.datadog.android.sample.R;
import com.datadog.android.sample.traces.TracesFragment;
import io.opentracing.Span;
import io.opentracing.util.GlobalTracer;

public class UserFragment extends Fragment implements View.OnClickListener {

    public static final String TAG_DIALOG = "user_dialog";

    public static final String PREF_ID = "user-id";
    public static final String PREF_NAME = "user-name";
    public static final String PREF_EMAIL = "user-email";


    private Logger mLogger = new Logger.Builder()
            .setLoggerName("user_fragment")
            .setLogcatLogsEnabled(true)
            .build();

    public static UserFragment newInstance() {
        return new UserFragment();
    }

    private EditText mId;
    private EditText mName;
    private EditText mEmail;
    private View mSave;

    // region Fragment

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_user, container, false);

        mId = rootView.findViewById(R.id.user_id);
        mName = rootView.findViewById(R.id.user_name);
        mEmail = rootView.findViewById(R.id.user_email);
        mSave = rootView.findViewById(R.id.save_user);

        mSave.setOnClickListener(this);
        return rootView;
    }

    @Override
    public void onResume() {
        super.onResume();

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());

        mId.setText(prefs.getString(PREF_ID, null));
        mName.setText(prefs.getString(PREF_NAME, null));
        mEmail.setText(prefs.getString(PREF_EMAIL, null));
    }

    // endregion

    // region View.OnClickListener

    @Override
    public void onClick(View v) {
        final Span span = GlobalTracer.get()
                .buildSpan("Saving User Info")
                .start();
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());

        String id = mId.getText().toString();
        String name = mName.getText().toString();
        String email = mEmail.getText().toString();

        prefs.edit()
                .putString(PREF_ID, id)
                .putString(PREF_NAME, name)
                .putString(PREF_EMAIL, email)
                .apply();

        Datadog.setUserInfo(id, name, email);
        span.log("Updated user info");
        span.finish();
    }

    // endregion

}

File: sample/kotlin/src/main/kotlin/com/datadog/android/sample/user/UserFragment.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.android.sample.user

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.fragment.app.Fragment
import com.datadog.android.Datadog
import com.datadog.android.sample.Preferences
import com.datadog.android.sample.R
import com.datadog.android.trace.logAttributes
import com.datadog.android.trace.withinSpan
import com.google.android.material.snackbar.Snackbar

internal class UserFragment : Fragment(), View.OnClickListener {

    lateinit var idField: EditText
    lateinit var nameField: EditText
    lateinit var emailField: EditText
    lateinit var userGenderField: EditText
    lateinit var userAgeField: EditText

    // region Fragment

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val rootView: View = inflater.inflate(R.layout.fragment_user, container, false)
        idField = rootView.findViewById(R.id.user_id)
        nameField = rootView.findViewById(R.id.user_name)
        emailField = rootView.findViewById(R.id.user_email)
        userGenderField = rootView.findViewById(R.id.user_gender)
        userAgeField = rootView.findViewById(R.id.user_age)
        rootView.findViewById<View>(R.id.save_user).setOnClickListener(this)
        return rootView
    }

    override fun onResume() {
        super.onResume()
        val preferences = Preferences.defaultPreferences(requireContext())
        idField.setText(preferences.getUserId())
        nameField.setText(preferences.getUserName())
        emailField.setText(preferences.getUserEmail())
        userGenderField.setText(preferences.getUserGender())
        userAgeField.setText(preferences.getUserAge().toString())
    }

    // endregion

    // region View.OnClickListener

    override fun onClick(v: View) {
        withinSpan("updateUserInfo") {
            val id: String = idField.text.toString()
            val name: String = nameField.text.toString()
            val email: String = emailField.text.toString()
            val gender: String = userGenderField.text.toString()
            val age: Int = Integer.valueOf(userAgeField.text.toString())
            Preferences.defaultPreferences(requireContext())
                .setUserCredentials(id, name, email, gender, age)
            Datadog.setUserInfo(id, name, email, emptyMap())
            Datadog.addUserProperties(
                mapOf<String, Any>(
                    GENDER_KEY to gender,
                    AGE_KEY to age
                )
            )
            logAttributes("Updated user info")
        }
        Snackbar.make(view ?: v.rootView, "User info updated", Snackbar.LENGTH_SHORT).show()
    }

    // endregion

    companion object {
        internal const val GENDER_KEY = "gender"
        internal const val AGE_KEY = "age"
    }
}
